src/docker-extract.c

```diff
@@ -0,0 +1,342 @@
+#define _GNU_SOURCE
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <errno.h>
+#include <string.h>
+#include <fcntl.h>
+#include <sys/file.h>
+#include <sys/stat.h>
+#include <archive.h>
+#include <archive_entry.h>
+
+#include "config.h"
+#include "util/file.h"
+#include "util/message.h"
+#include "util/registry.h"
+#include "util/util.h"
+
+/* apply_opaque
+ *  Given opq_marker as a path to a whiteout opaque marker
+ *    e.g. usr/share/doc/test/.wh..wh..opq
+ *  Make the directory containing the make opaque for this layer by removing it
+ *  if it exists under rootfs_dir
+ */
+int apply_opaque(const char *opq_marker, char *rootfs_dir) {
+    int retval = 0;
+    char *token, *opq_dir, *opq_dir_rootfs;
+    size_t buff_len;
+
+    token = strrchr(opq_marker, '/');
+
+    if (token == NULL) {
+        singularity_message(ERROR, "Error getting dirname for opaque marker\n");
+        ABORT(255);
+    }
+
+    buff_len = token - opq_marker + 1;
+    opq_dir = malloc(buff_len);
+    snprintf(opq_dir, buff_len, "%s", opq_marker);
+
+    buff_len = strlen(rootfs_dir) + 1 + strlen(opq_dir) + 1;
+    opq_dir_rootfs = malloc(buff_len);
+    snprintf(opq_dir_rootfs, buff_len, "%s/%s", rootfs_dir, opq_dir);
+
+    if (is_dir(opq_dir_rootfs) == 0) {
+        s_rmdir(opq_dir_rootfs);
+    }
+
+    free(opq_dir);
+    free(opq_dir_rootfs);
+
+    return retval;
+}
+
+/* apply_whiteout
+ *  Given wh_marker as a path to a whiteout marker
+ *    e.g. usr/share/doc/test/.wh.deletedfile
+ *  Whiteout the referenced file for this layer by removing it if it exists
+ *  under rootfs_dir
+ */
```

Comment:
For this function also, manipulating strings with snprintf and using buffers, you can use the example I wrote to make it more simple.

Reply:
Hopefully c8baf0d is something more like it for this too.